src/main/java/servico/bookstore/Bookstore.java

```diff
@@ -703,66 +705,78 @@ public Stock getStock(int bId) {
+
+        return clientIds;
+    }
+
+    private Map<Integer, BookstoreBookCounter> countBooksBoughtByClients(Set<Integer> clientIds, Book targetBook) {
+        Map<Integer, BookstoreBookCounter> counters = new HashMap<>();
+
+        // Para cada ordem, se o cliente comprou o livro alvo, conta os outros livros comprados
+        ordersByCreation.stream()
+            .filter(order -> clientIds.contains(order.getCustomer().getId()))
+            .forEach(order -> {
+                order.getLines().forEach(line -> {
                     Book book = line.getBook();
-                    if (targetBook.getId() != book.getId()) {
-                        Counter counter = counters.get(book.getId());
-                        if (counter == null) {
-                            counter = new Counter();
-                            counter.book = book;
-                            counter.count = 0;
-                            counters.put(book.getId(), counter);
-                        }
-                        counter.count += line.getQty();
+                    if (book.getId() != targetBook.getId()) {
+                        counters.computeIfAbsent(book.getId(), id -> new BookstoreBookCounter(book))
+                                .addQuantity(line.getQty());
                     }
                 });
-            }
-        }
-        Counter[] sorted = counters.values().toArray(new Counter[] {});
-        Arrays.sort(sorted, (Counter a, Counter b) -> {
-            if (b.count > a.count) {
-                return 1;
-            }
-            return b.count < a.count ? -1 : 0;
-        });
-        Book[] related = new Book[] { targetBook, targetBook, targetBook,
-                targetBook, targetBook };
-        for (j = 0; j < 5 && j < sorted.length; j++) {
-            related[j] = sorted[j].book;
+            });
+
+        return counters;
+    }
+
+    private Book[] getTopFiveRelatedBooks(Map<Integer, BookstoreBookCounter> bookCounters, Book targetBook) {
+        // Ordena os livros mais vendidos e pega os cinco primeiros
+        List<BookstoreBookCounter> sortedCounters = bookCounters.values().stream()
+            .sorted(Comparator.comparingInt(BookstoreBookCounter::getCount).reversed())
+            .collect(Collectors.toList());
+
+        // Preenche com o livro alvo, caso existam menos de 5 livros relacionados
+        Book[] relatedBooks = new Book[5];
+        for (int i = 0; i < Math.min(5, sortedCounters.size()); i++) {
+            relatedBooks[i] = sortedCounters.get(i).getBook();
         }
-        targetBook.setRelated1(related[0]);
-        targetBook.setRelated2(related[1]);
-        targetBook.setRelated3(related[2]);
-        targetBook.setRelated4(related[3]);
-        targetBook.setRelated5(related[4]);
+        Arrays.fill(relatedBooks, relatedBooks.length, 5, targetBook);
+
+        return relatedBooks;
+    }
+
+    private void setRelatedBooks(Book targetBook, Book[] relatedBooks) {

```

Comment:
```suggestion

    /**
     * Sets the five related book references in the target book.
     *
     * @param targetBook the book to update with related recommendations
     * @param relatedBooks array of five books to set as related items
     */
    private void setRelatedBooks(Book targetBook, Book[] relatedBooks) {

```

adicionando a porra do javadoc